### PR TITLE
Fix program by slug when BYPASS_LOGIN_LANGUAGE_SCREENS is on.

### LINKED ISSUE #4856: Program Deep Link Throws If Not Already Logged In
**Describe the bug**
This started after the switch to enable BYPASS_LOGIN_LANGUAGE_SCREENS. When a user loads civiform from a program deep link an exception is occurring if they are not already logged in as a guest or with an account. 

Prior to the new login format the redirect controller would handle this by going to the login page. I think [that this code](https://github.com/civiform/civiform/blob/main/server/app/controllers/applicant/RedirectController.java#L80-L83) would still work if there is a new controller action to redirect to, but I don't know which that is.

**To Reproduce**

1. Open a new private browser window that has not been to https://civiform.seattle.gov
2. In the browser address bar paste https://civiform.seattle.gov/programs/upcoming and hit enter to load the page
3. Receive an exception page (See below for details)

For local development just create and publish a program first and use that deep link. 

**Workaround**

1. Go to https://civiform.seattle.gov, this will trigger a guest account
2. In the browser address bar paste https://civiform.seattle.gov/programs/upcoming and hit enter to load the page
3. Page will load

**Expected behavior**
The deep link should load whether or not the client was already logged in as guest or with a user account.

**Exception**
```
@81hn0ahc9 - Internal server error, for (GET) [/programs/upcoming]

play.api.UnexpectedException: Unexpected exception[NoSuchElementException: No value present]
 at play.api.http.HttpErrorHandlerExceptions$.throwableToUsefulException(HttpErrorHandler.scala:358)
 at play.api.http.HttpErrorHandlerExceptions.throwableToUsefulException(HttpErrorHandler.scala)
 at play.http.DefaultHttpErrorHandler.throwableToUsefulException(DefaultHttpErrorHandler.java:246)
 at play.http.DefaultHttpErrorHandler.onServerError(DefaultHttpErrorHandler.java:182)
 at controllers.ErrorHandler.onServerError(ErrorHandler.java:102)
 at play.core.j.JavaHttpErrorHandlerAdapter.onServerError(JavaHttpErrorHandlerAdapter.scala:32)
 at play.core.server.AkkaHttpServer$$anonfun$2.applyOrElse(AkkaHttpServer.scala:441)
 at play.core.server.AkkaHttpServer$$anonfun$2.applyOrElse(AkkaHttpServer.scala:433)
 at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:490)
 at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:63)
 at akka.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:100)
 at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
 at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:94)
 at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:100)
 at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:49)
 at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:48)
 at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
 at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
 at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
 at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)

Caused by: java.util.NoSuchElementException: No value present
 at java.base/java.util.Optional.get(Unknown Source)
 at controllers.applicant.RedirectController.programBySlug(RedirectController.java:89)
 at router.Routes$$anonfun$routes$1.$anonfun$applyOrElse$303(Routes.scala:3622)
 at play.core.routing.HandlerInvokerFactory$$anon$9.resultCall(HandlerInvoker.scala:154)
 at play.core.routing.HandlerInvokerFactory$JavaActionInvokerFactory$$anon$3$$anon$4$$anon$5.invocation(HandlerInvoker.scala:115)
 at play.core.j.JavaAction$$anon$1.call(JavaAction.scala:119)
 at play.http.DefaultActionCreator$1.call(DefaultActionCreator.java:33)
 at play.core.j.JavaAction.$anonfun$apply$8(JavaAction.scala:175)
 at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:678)
 at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:467)
 at play.core.j.HttpExecutionContext.$anonfun$execute$1(HttpExecutionContext.scala:64)
 at play.api.libs.streams.Execution$trampoline$.executeScheduled(Execution.scala:109)
 at play.api.libs.streams.Execution$trampoline$.execute(Execution.scala:71)
 at scala.concurrent.impl.Promise$Transformation.submitWithValue(Promise.scala:429)
 at scala.concurrent.impl.Promise$DefaultPromise.submitWithValue(Promise.scala:338)
 at scala.concurrent.impl.Promise$DefaultPromise.dispatchOrAddCallbacks(Promise.scala:312)
 at scala.concurrent.impl.Promise$DefaultPromise.flatMap(Promise.scala:176)
 at play.api.libs.streams.Accumulator$.$anonfun$futureToSink$1(Accumulator.scala:231)
 at akka.stream.impl.Transform.apply(TraversalBuilder.scala:159)
 at akka.stream.impl.PhasedFusingActorMaterializer.materialize(PhasedFusingActorMaterializer.scala:522)
 at play.core.server.AkkaHttpServer$$anonfun$2.applyOrElse(AkkaHttpServer.scala:441)
 at play.core.server.AkkaHttpServer$$anonfun$2.applyOrElse(AkkaHttpServer.scala:433)
 at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:490)
 at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:63)
 at akka.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:100)
 at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
 at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:94)
 at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:100)
 at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:49)
 at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:48)
 at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
 at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
 at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
 at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)

Caused by: java.util.NoSuchElementException: No value present
 at java.base/java.util.Optional.get(Unknown Source)
 at controllers.applicant.RedirectController.programBySlug(RedirectController.java:89)
 at router.Routes$$anonfun$routes$1.$anonfun$applyOrElse$303(Routes.scala:3622)
 at play.core.routing.HandlerInvokerFactory$$anon$9.resultCall(HandlerInvoker.scala:154)
 at play.core.routing.HandlerInvokerFactory$JavaActionInvokerFactory$$anon$3$$anon$4$$anon$5.invocation(HandlerInvoker.scala:115)
 at play.core.j.JavaAction$$anon$1.call(JavaAction.scala:119)
 at play.http.DefaultActionCreator$1.call(DefaultActionCreator.java:33)
 at play.core.j.JavaAction.$anonfun$apply$8(JavaAction.scala:175)
 at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:678)
 at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:467)
 at play.core.j.HttpExecutionContext.$anonfun$execute$1(HttpExecutionContext.scala:64)
 at play.api.libs.streams.Execution$trampoline$.executeScheduled(Execution.scala:109)
 at play.api.libs.streams.Execution$trampoline$.execute(Execution.scala:71)
 at scala.concurrent.impl.Promise$Transformation.submitWithValue(Promise.scala:429)
 at scala.concurrent.impl.Promise$DefaultPromise.submitWithValue(Promise.scala:338)
 at scala.concurrent.impl.Promise$DefaultPromise.dispatchOrAddCallbacks(Promise.scala:312)
 at scala.concurrent.impl.Promise$DefaultPromise.flatMap(Promise.scala:176)
 at play.api.libs.streams.Accumulator$.$anonfun$futureToSink$1(Accumulator.scala:231)
 at akka.stream.impl.Transform.apply(TraversalBuilder.scala:159)
 at akka.stream.impl.PhasedFusingActorMaterializer.materialize(PhasedFusingActorMaterializer.scala:522)
```

### PR DESCRIPTION
**Pull request by michaelzetune**
### Description

Please explain the changes you made here.

## Release notes

The title of the pull request will be used as the default release notes description. If more detail is needed to communicate to partners the scope of the PR's changes, use this release notes section.

#### General

- [x] Added the correct label: < feature | bug | dependencies | infrastructure | ignore-for-release >
- [x] Created tests which fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

#### User visible changes

- [ ] Wrote browser tests using the [validateAccessibility](https://sourcegraph.com/github.com/civiform/civiform/-/blob/browser-test/src/support/index.ts?L437:14&subtree=true) method
- [ ] Tested on mobile view. See [mobile device mode](https://developer.chrome.com/docs/devtools/device-mode/)
- [ ] Manually tested at 200% size
- [ ] Manually evaluated tab order

#### New Features

- [ ] Add new FeatureFlag env vars to `server/conf/application.conf`
- [ ] Conditioned new functionality on a [FeatureFlag](https://docs.civiform.us/contributor-guide/developer-guide/feature-flags)
- [ ] Wrote browser tests with the feature flag off and on, etc.

### Issue(s) this completes

Fixes #4856

### CODE REVIEW

File: browser-test/src/support/index.ts [hunk 1]
```diff
@@ -183,55 +185,18 @@ export interface TestContext {
  *     so none of its properties should be cached and reused between tests.
  */
 export const createTestContext = (clearDb = true): TestContext => {
-  let browser: Browser
-  let browserContext: BrowserContext
+  // let browser: Browser
+  // let browserContext: BrowserContext
```
Comment (gwendolyngoetz) [inline browser-test/src/support/index.ts:189]: Please delete the commented out code if obsolete.

File: browser-test/src/program_deep_link.test.ts [hunk 2]
```diff
@@ -5,7 +5,7 @@ import {
   loginAsAdmin,
   loginAsGuest,
   loginAsTestUser,
-  logout,
+  logout, resetContext,
```
Comment (gwendolyngoetz) [inline browser-test/src/program_deep_link.test.ts:8]: Please put resetContext on a new line. /bin/fmt will probably fix it automatically.

**Review by gwendolyngoetz — CHANGES_REQUESTED** [review]

File: browser-test/src/program_deep_link.test.ts [hunk 3]
```diff
@@ -96,4 +101,25 @@ describe('navigating to a deep link', () => {
...
   })
+
+  it('Non-logged in user should get redirected to the program page and not an error', async () => {
+    await resetContext(ctx)
+    const {page, browserContext} = ctx
+
+    await selectApplicantLanguage(page, 'English')
+
+    await logout(page)
+    await browserContext.clearCookies()
```
Comment (michaelzetune) [inline browser-test/src/program_deep_link.test.ts:112]: if `clearCookies()` accomplishes what `resetContext()` was meant to do here, then maybe it's worth simplifying the changes in `index.ts`

File: server/app/controllers/applicant/RedirectController.java [hunk 4]
```diff
@@ -77,11 +78,14 @@ public RedirectController(
...
-    if (!featureFlags.getFlagEnabled(request, FeatureFlag.BYPASS_LOGIN_LANGUAGE_SCREENS)
-        && profile.isEmpty()) {
-      Result result = redirect(routes.HomeController.loginForm(Optional.of("login")));
+    if (profile.isEmpty()) {
+      Result result;
+      if (featureFlags.getFlagEnabled(request, FeatureFlag.BYPASS_LOGIN_LANGUAGE_SCREENS)) {
+        result = redirect(routes.CallbackController.callback(GuestClient.CLIENT_NAME).url());
+      } else {
+        result = redirect(routes.HomeController.loginForm(Optional.of("login")));
+      }
```
Comment (bion) [inline server/app/controllers/applicant/RedirectController.java:87]: **nitpick (non-blocking):** personal preference would be a ternary here but nbd

**Review by bion — APPROVED** [review]

File: browser-test/src/program_deep_link.test.ts (hunk 3 above)
Comment (gwendolyngoetz) [reply browser-test/src/program_deep_link.test.ts:112]: I tried putting resetContext back and it did not pass. The call to clearCookies was just meant to be double certain there were no lingering auth cookies around. I can probably remove it as the tests still pass without.

**Review by gwendolyngoetz — APPROVED** [review]

**Pull request merged**